### ISSUE TITLE
Texture compression for 3D Tiles and glTF

### ISSUE THREAD

**Issue by pjcozzi** [AnalyticalGraphicsInc/cesium#4679]
@bagnell @lilleyse please add comments with your latest notes.

@sbtron we will be very happy to collaborate on glTF texture compression (see linked glTF issues below), but probably need a few more weeks of research.

* [x] Quickly pull together a proof-of-concept pipeline that compresses textures, stores them in a glTF `extras` properties, and loads them in Cesium. #4758
* [x] Figure what formats to support (we’ll need multiple formats to cover mobile and desktop) and what compression tools work well for our use cases.
   * WebGL extensions: [S3TC](https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_s3tc/), [ATC](https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_atc/) (likely not worth it), [PVRTC](https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_pvrtc/), [ETC1](https://www.khronos.org/registry/webgl/extensions/WEBGL_compressed_texture_etc1/), ETC2 (WebGL 2)
   * PlayCanvas [does](https://blog.playcanvas.com/webgl-texture-compression-made-easy/) DXT (desktop), PVRTC (iOS), and ETC1 (Android)
* [x] Flush out glTF extension spec.  What is the container texture file format?
   * [x] Reference in `extensionsUsed`/`extensionsRequired` (glTF 1.1)
   * [x] In glTF 1.1, will need to list the GL extensions in `glExtensionsUsed`; however, we won't want to have, for example, three copies of a glTF file to support each referenced .ktx file.
   * Improving texture support for 1.0.1+, https://github.com/KhronoxsGroup/glTF/issues/620
   * Evaluating texture storage for different APIs, https://github.com/KhronosGroup/glTF/issues/739
* [x] Flush out Cesium: 3D Tiles and glTF loader.
* [x] Flush out glTF-pipeline texture compressor(s) integration (see below).
* [x] Tech post for the Cesium blog.

**Tools**

* Texture compression plugin point for glTF-pipeline.  Tools to consider (the first three are likely the strongest):
   * Google’s ETC2Comp.  [code](https://github.com/google/etc2comp), [blog post](https://opensource.googleblog.com/2016/11/etc2comp-fast-texture-compression.html), [more detailed blog post](https://medium.com/@duhroach/building-a-blazing-fast-etc2-compressor-307f3e9aad99#.rrkx8irvn).  Apache 2.0 license.
   * AMD Compressonator. [code](https://github.com/GPUOpen-Tools/Compressonator), [website](http://gpuopen.com/gaming-product/compressonator/).  MIT license.
   * [Transcode](http://www.binomial.info/blog/2016/6/22/introduction-to-gpu-texture-transcoding) JPG-like format to DXT using [Crunch](https://github.com/BinomialLLC/crunch).  ZLIB license.  [JavaScript loading](https://plus.google.com/+BrandonJonesToji/posts/KJ42NGorLTj).  ([used](https://docs.unity3d.com/Manual/webgl-building.html) by Unity)
   * NVIDIA Texture Tools (DXT / BCn Compressor)
   * Imagination’s [PVRTexTool](https://community.imgtec.com/developers/powervr/tools/pvrtextool/).  Not open source?
   * [astc-encoder](https://github.com/ARM-software/astc-encoder) Developed by ARM. Used in PVRTexTool. [License](https://github.com/ARM-software/astc-encoder/blob/master/license.txt)
   * [FasTC](https://github.com/GammaUNC/FasTC).  Apache 2.0 license.
   * [Etcpak](http://richg42.blogspot.com/2016/09/etcpak.html).  Open source, license TBA.
   * Apple’s [texturetool](https://developer.apple.com/library/content/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/TextureTool/TextureTool.html#//apple_ref/doc/uid/TP40008793-CH108-SW1).  Not open source?
   * [Fast ISPC Texture Compressor](https://github.com/GameTechDev/ISPCTextureCompressor) 

**Longer-term**

* Compress leaf tile textures?  Different compression rate at different tree levels?
* Use this for Cesium imagery.
* When can we use [GST: GPU-decodable Supercompressed Textures](http://richg42.blogspot.com/2016/10/gst-gpu-decodable-supercompressed.html)?

**Misc notes**

* DXT: 2-10x faster than ASTC. ASTC 4-20x faster than ETC2
* Tegra is only mobile GPU with DXT
* PVRTC/PVRTC2 - PowerVR proprietary - iOS.  Square textures only says the Intel article?!?!?!
* ETC/ETC2 - Khronos ES 2.0 extension - no alpha in ETC1 - better quality than DXT
* ASTC - Khronos ES 3.0 - wide range of compression ratios. adaptive per 128-bit block.  Android Extension Pack.
* zip/lsw - better than jpg?!?!!?

**Resources**

* What texture compression formats does my system support? http://toji.github.io/texture-tester/
* [Texture Compression](https://docs.google.com/presentation/d/1qwkxbpGSkFHlrcR3mcc2pHyApKVxny441QYbx4z72NM/edit#slide=id.p) at CppCon
* [Texture formats for 2D games, part 4: Overview](http://joostdevblog.blogspot.nl/2015/12/texture-formats-for-2d-games-part-4.html)
* Old: [Compressed Textures in WebGL](http://blog.tojicode.com/2011/12/compressed-textures-in-webgl.html)
* [Using ASTC Texture Compression for Game Assets](https://developer.nvidia.com/astc-texture-compression-for-game-assets)
* [Android* Texture Compression - a comparison study with code sample](https://software.intel.com/en-us/articles/android-texture-compression-a-comparison-study-with-code-sample?language=de)
* [Efficient Game Textures with Hardware Compression](http://android-developers.blogspot.com/2015/01/efficient-game-textures-with-hardware.html)

**Comment by pjcozzi** [comment]
Initial numbers for a texture from a reality model via @bagnell:

```
Raw size         3MB    // Also the size in GPU memory without compression
PNG              928KB
JPG              229KB
 
For desktop:
DXT3             1MB    // Size in GPU memory with compression.  Note that DXT3 is RGBA.
DXT3 + gzip      238KB  // Wow – size over the wire is on par with jpg!
 
For iOS:
PVR 4bpp         513KB
PVR 4bpp + gzip  213KB  // On par with jpg
PVR 2bpp         257KB
PVR 2bpp + gzip  114KB  // Better than jpg, but check quality
 
For Android:
ETC1             513KB
ETC1 + gzip      217KB
```

**Comment by pjcozzi** [comment]
There's a JCGT paper on FasTC: http://jcgt.org/published/0003/04/07/

**Comment by pjcozzi** [comment]
* Crunch/Emscripten performance numbers (old): http://www-cs-students.stanford.edu/~eparker/files/crunch/decode_test.html
   * When texture compression is not supported, they transcode DXT to 565.  This could be a good option so we don't have an extra set of uncompressed textures to serve. http://www-cs-students.stanford.edu/~eparker/files/crunch/dxt-to-rgb565.js

**Comment by pjcozzi** [comment]
Fixed in #4758 and https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/204.

There will be more work for us to expose this through the model converter, do the official glTF spec, and eventually adopt the universal format.
